File: genkit-tools/common/src/types/evaluator.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * This file defines schema and types that are used by Genkit evaluators.
 *
 * NOTE: Keep this file in sync with genkit/ai/src/evaluator.ts!
 * Eventually tools will be source of truth for these types (by generating a
 * JSON schema) but until then this file must be manually kept in sync
 */
import { z } from 'zod';

//
// IMPORTANT: Keep this file in sync with genkit/ai/src/evaluator.ts!
//

/**
 * Zod schema for a data point for evaluation
 */
export const BaseDataPointSchema = z.object({
  input: z.unknown(),
  output: z.unknown().optional(),
  context: z.array(z.unknown()).optional(),
  reference: z.unknown().optional(),
  testCaseId: z.string().optional(),
  traceIds: z.array(z.string()).optional(),
});
export type BaseDataPoint = z.infer<typeof BaseDataPointSchema>;

// DataPoint that is to be used for actions. This needs testCaseId to be
// present.
export const BaseEvalDataPointSchema = BaseDataPointSchema.extend({
  testCaseId: z.string(),
});
export type BaseEvalDataPoint = z.infer<typeof BaseEvalDataPointSchema>;
// Enum for Score Status
export const EvalStatusEnumSchema = z.enum(['UNKNOWN', 'PASS', 'FAIL']);
/**
 * Zod schema for evaluation score
 */
export const ScoreSchema = z.object({
  id: z
    .string()
    .describe('Optional ID to differentiate different scores')
    .optional(),
  score: z.union([z.number(), z.string(), z.boolean()]).optional(),
  status: EvalStatusEnumSchema.optional(),
  error: z.string().optional(),
  details: z
    .object({
      reasoning: z.string().optional(),
    })
    .passthrough()
    .optional(),
});
export type Score = z.infer<typeof ScoreSchema>;

/**
 * Zod schema for an evaluator action request.
 */
export const EvalRequestSchema = z.object({
  dataset: z.array(BaseDataPointSchema),
  evalRunId: z.string(),
  options: z.unknown(),
});
export type EvalRequest = z.infer<typeof EvalRequestSchema>;

/**
 * Zod schema for single response from evaluator function.
 *
 * Disclaimer, this schema is defined as `EvalResponseSchema` in
 * genkit/js/ai/evaluator.ts. That is a misnomer and this is only the response
 * from the user-provided callback function, we rename this accordingly here.
 */
export const EvalFnResponseSchema = z.object({
  sampleIndex: z.number().optional(),
  testCaseId: z.string(),
  traceId: z.string().optional(),
  spanId: z.string().optional(),
  evaluation: z.union([ScoreSchema, z.array(ScoreSchema)]),
});
export type EvalFnResponse = z.infer<typeof EvalFnResponseSchema>;

/**
 * Zod schema for batch evaluator responses, typically used as the evaluator
 * response for a dataset.
 *
 * Disclaimer, this schema is defined as `EvalResponsesSchema` in
 * genkit/js/ai/evaluator.ts. That is not an ideal name to differentiate from
 * `EvalResponseSchema`, so we rename this to EvalResponse, given that the
 * former now has a new name in this package.
 */
export const EvalResponseSchema = z.array(EvalFnResponseSchema);
export type EvalResponse = z.infer<typeof EvalResponseSchema>;

File: py/packages/genkit/src/genkit/blocks/evaluator.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Evaluator type definitions for the Genkit framework.

This module defines the type interfaces for evaluators in the Genkit framework.
Evaluators are used for assessint the quality of output of a Genkit flow or
model.
"""

from collections.abc import Callable, Coroutine
from typing import Any, ClassVar, TypeVar, cast

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

from genkit.core.action import ActionMetadata
from genkit.core.action.types import ActionKind
from genkit.core.schema import to_json_schema
from genkit.core.typing import (
    BaseDataPoint,
    EvalFnResponse,
    EvalRequest,
)

T = TypeVar('T')

# User-provided evaluator function that evaluates a single datapoint.
# Must be async (coroutine function).
EvaluatorFn = Callable[[BaseDataPoint, T], Coroutine[Any, Any, EvalFnResponse]]

# User-provided batch evaluator function that evaluates an EvaluationRequest
BatchEvaluatorFn = Callable[[EvalRequest, T], Coroutine[Any, Any, list[EvalFnResponse]]]


class EvaluatorRef(BaseModel):
    """Reference to an evaluator."""

    model_config: ClassVar[ConfigDict] = ConfigDict(extra='forbid', populate_by_name=True, alias_generator=to_camel)

    name: str
    config_schema: dict[str, object] | None = None


def evaluator_ref(name: str, config_schema: dict[str, object] | None = None) -> EvaluatorRef:
    """Create a reference to an evaluator.

    Args:
        name: Name of the evaluator.
        config_schema: Optional schema for evaluator configuration.

    Returns:
        An EvaluatorRef instance.
    """
    return EvaluatorRef(name=name, config_schema=config_schema)


def evaluator_action_metadata(
    name: str,
    config_schema: type | dict[str, Any] | None = None,
) -> ActionMetadata:
    """Generates an ActionMetadata for evaluators.

    Args:
        name: Name of the evaluator.
        config_schema: Optional schema for evaluator configuration.

    Returns:
        An ActionMetadata instance for the evaluator.
    """
    return ActionMetadata(
        kind=cast(ActionKind, ActionKind.EVALUATOR),
        name=name,
        input_json_schema=to_json_schema(EvalRequest),
        output_json_schema=to_json_schema(list[EvalFnResponse]),
        metadata={'evaluator': {'customOptions': to_json_schema(config_schema) if config_schema else None}},
    )
